src/v/cloud_storage/cache_service.cc

```diff
@@ -302,199 +302,305 @@ ss::future<> cache::trim() {
-                vlog(
-                  cst_log.error,
-                  "Cache eviction couldn't delete {}: {}.",
-                  file_stat.path,
-                  e.what());
-            }
-            candidate_i++;
+            vlog(
+              cst_log.trace,
+              "trim: reclaimed(fast) {} bytes from {}",
+              this_segment_deleted_bytes,
+              file_stat.path);
+        } catch (const ss::gate_closed_exception&) {
+            // We are shutting down, stop iterating and propagate
+            throw;
+        } catch (const std::exception& e) {
+            vlog(
+              cst_log.error,
+              "trim: couldn't delete {}: {}.",
+              file_stat.path,
+              e.what());
         }
+        candidate_i++;
+    }
 
-        // We aim to keep current_cache_size continuously up to date, but
-        // in case of housekeeping issues, correct it if it apepars to have
-        // drifted too far from the result of our directory walk.
-        // This is a lower bound that permits current cache size to deviate
-        // by the amount of data currently in tmp files, because they may be
-        // updated while the walk is happening.
-        uint64_t cache_size_lower_bound = walked_cache_size - deleted_size
-                                          - tmp_files_size;
-        if (_current_cache_size < cache_size_lower_bound) {
-            vlog(
-              cst_log.debug,
-              "Correcting cache size drift ({} -> {})",
-              _current_cache_size,
-              cache_size_lower_bound);
-            _current_cache_size = cache_size_lower_bound;
-            _current_cache_objects = filtered_out_files
-                                     + candidates_for_deletion.size()
-                                     - deleted_count;
+    co_return result;
+}
+
+ss::future<cache::trim_result>
+cache::trim_exhaustive(uint64_t size_to_delete, size_t objects_to_delete) {
+    trim_result result;
+
+    // Enumerate ALL files in the cache (as opposed to trim_fast that strips out
+    // indices/tx/tmp files)
+    auto [walked_cache_size, _filtered_out, candidates, _]
+      = co_await _walker.walk(_cache_dir.native(), _access_time_tracker);
+
+    // Sort by atime
+    std::sort(candidates.begin(), candidates.end(), [](auto& a, auto& b) {
+        return a.access_time < b.access_time;
+    });
+
+    size_t candidate_i = 0;
+    while (
+      candidate_i < candidates.size()
+      && (result.deleted_size < size_to_delete || result.deleted_count < objects_to_delete)) {
+        auto& file_stat = candidates[candidate_i++];
+
+        // Skip the accesstime file, we should never delete this.
+        if (
+          file_stat.path
+            == (_cache_dir / access_time_tracker_file_name).string()
+          || file_stat.path
+               == (_cache_dir / access_time_tracker_file_name_tmp).string()) {
+            candidate_i++;
+            continue;
```

Comment:
This should be a helper as it is duplicated in both trim methods now.

Reply:
I'll do a quick followup for that.